File: Fraction.vb
Imports System.Numerics
Public Structure Fraction
    Implements IComparable, IComparable(Of Fraction), IEquatable(Of Fraction)
    Private _Numerator As BigInteger
    Private _Denominator As BigInteger
    Public Property Numerator As BigInteger
        Get
            Return _Numerator
        End Get
        Set(value As BigInteger)
            _Numerator = value
            Simplify()
        End Set
    End Property
    Public Property Denominator As BigInteger
        Get
            Return _Denominator
        End Get
        Set(value As BigInteger)
            _Denominator = value
            Simplify()
        End Set
    End Property
    Public Property WholeNumber As BigInteger
        Get
            If Denominator = 0 Then Return 0
            Return BigInteger.Divide(Numerator, Denominator)
        End Get
        Set(value As BigInteger)
            Dim ProperFraction = Me.ProperFraction
            Dim Fraction = ProperFraction + value
            _Numerator = Fraction.Numerator
            _Denominator = Fraction.Denominator
            Simplify()
        End Set
    End Property
    Public Property ProperFraction As Fraction
        Get
            If Denominator = 0 Then Return New Fraction(0, 0)
            Return New Fraction(Numerator Mod Denominator, Denominator)
        End Get
        Set(value As Fraction)
            Dim WholeNumber = Me.WholeNumber
            Dim Fraction = WholeNumber + value
            _Numerator = Fraction.Numerator
            _Denominator = Fraction.Denominator
            Simplify()
        End Set
    End Property
    Public ReadOnly Property IsProper As Boolean
        Get
            If BigInteger.Abs(Numerator) < Denominator Then Return True
            Return False
        End Get
    End Property
    Public ReadOnly Property IsImproper As Boolean
        Get
            Return Not IsProper
        End Get
    End Property
    Public ReadOnly Property IsNull As Boolean
        Get
            If Numerator = 0 Then Return True
            Return False
        End Get
    End Property
    Public ReadOnly Property IsUnit As Boolean
        Get
            If Numerator = 1 And Denominator = 1 Then Return True
            Return False
        End Get
    End Property
    Public ReadOnly Property IsNegitive As Boolean
        Get
            If Numerator < 0 Then Return True
            Return False
        End Get
    End Property
    Public ReadOnly Property IsUndefined As Boolean
        Get
            If Denominator = 0 Then Return True
            Return False
        End Get
    End Property
    Public ReadOnly Property Absolute As Fraction
        Get
            Return New Fraction(BigInteger.Abs(Numerator), BigInteger.Abs(Denominator))
        End Get
    End Property

    Public Shadows Function ToString() As String
        Return _Numerator.ToString + "/" + _Denominator.ToString
    End Function
    Public Shadows Function ToString(DigitsAfterPoint As ULong) As String
        Dim Whole As BigInteger = BigInteger.Abs(WholeNumber)
        Dim Proper As Fraction = ProperFraction.Absolute
        Dim Value As String = Whole.ToString + "."
        Dim Numerator As BigInteger = Proper.Numerator
        Dim Denominator As BigInteger = Proper.Denominator
        For i = 0 To DigitsAfterPoint - 1
            Numerator = Numerator * 10
            Value += BigInteger.Divide(Numerator, Denominator).ToString
            Numerator = Numerator Mod Denominator
        Next
        If Me.IsNegitive Then
            Return "-" + Value
        Else
            Return Value
        End If
    End Function

    Public Sub SetValue(Numerator As BigInteger, Denominator As BigInteger)
        Me._Numerator = Numerator
        Me._Denominator = Denominator
        Me.Simplify()
    End Sub
    Public Sub Simplify()
        If Denominator = 0 Then
            _Numerator = 0
            _Denominator = 0
        ElseIf Denominator < 0 Then
            _Denominator = BigInteger.Abs(Denominator)
            _Numerator = -Numerator
        ElseIf Numerator = 0 And Denominator <> 0 Then
            _Denominator = 1
        ElseIf Numerator = Denominator And Denominator <> 0 Then
            _Numerator = 1
            _Denominator = 1
        Else
            Dim GDC = Fraction.GCD(Numerator, Denominator)
            Me._Numerator = Numerator / GDC
            Me._Denominator = Denominator / GDC
        End If
    End Sub
    Public Sub Reduce(Optional Change As Long = 1)
        Me.Reduce(Change, Change)
    End Sub
    Public Sub Reduce(ChangeInNumerator As Long, ChangeInDenominator As Long)
        Dim IsNegitive As Boolean = False
        Dim Proper As Fraction = Me.ProperFraction
        If Proper < 0 Then
            IsNegitive = True
            Proper = Proper.Absolute
        End If
        Dim Reduced As Fraction = Proper
        For CIN As Long = -ChangeInNumerator To ChangeInNumerator
            For CID As Long = -ChangeInDenominator To ChangeInDenominator
                Dim Fraction = (New Fraction(Proper.Numerator + CIN, Proper.Denominator + CID)).Absolute
                If _
                    Fraction.Numerator < Reduced.Numerator And
                    Fraction.Denominator < Reduced.Denominator And
                    Not Fraction.IsUndefined Then Reduced = Fraction
            Next
        Next
        If IsNegitive = True Then
            Me.ProperFraction = -Reduced
        Else
            Me.ProperFraction = Reduced
        End If
    End Sub

    Public Function CompareTo(obj As Object) As Integer Implements IComparable.CompareTo
        Return Me.CompareTo(obj)
    End Function
    Public Function CompareTo(other As Fraction) As Integer Implements IComparable(Of Fraction).CompareTo
        Return Me.WholeNumber - other.WholeNumber
    End Function
    Public Overloads Function Equals(other As Fraction) As Boolean Implements IEquatable(Of Fraction).Equals
        If Me = other Then Return True
        Return False
    End Function

    Sub New(Numerator As BigInteger, Denominator As BigInteger)
        Me._Numerator = Numerator
        Me._Denominator = Denominator
        Simplify()
    End Sub
    Sub New(Number As BigInteger)
        Me._Numerator = Number
        Me._Denominator = 1
    End Sub
    Sub New(WholeNumber As BigInteger, ProperFraction As Fraction)
        Me._Numerator = WholeNumber * ProperFraction.Denominator + ProperFraction.Numerator
        Me._Denominator = ProperFraction.Denominator
    End Sub
    Sub New(Value As String)
        Dim Fraction As Fraction = Fraction.Parse(Value)
        Me._Numerator = Fraction.Numerator
        Me._Denominator = Fraction.Denominator
        Simplify()
    End Sub
    Sub New(Number As Decimal)
        Me.New(Number.ToString)
    End Sub
    Sub New(Number As Double)
        Me.New(Number.ToString)
    End Sub
    Sub New(Number As Single)
        Me.New(Number.ToString)
    End Sub

    Public Shared Operator +(Fraction1 As Fraction, Fraction2 As Fraction) As Fraction
        Return New Fraction(BigInteger.Add(BigInteger.Multiply(Fraction1.Numerator, Fraction2.Denominator), BigInteger.Multiply(Fraction2.Numerator, Fraction1.Denominator)), BigInteger.Multiply(Fraction1.Denominator, Fraction2.Denominator))
    End Operator
    Public Shared Operator -(Fraction1 As Fraction, Fraction2 As Fraction) As Fraction
        Return New Fraction(BigInteger.Subtract(BigInteger.Multiply(Fraction1.Numerator, Fraction2.Denominator), BigInteger.Multiply(Fraction2.Numerator, Fraction1.Denominator)), BigInteger.Multiply(Fraction1.Denominator, Fraction2.Denominator))
    End Operator
    Public Shared Operator *(Fraction1 As Fraction, Fraction2 As Fraction) As Fraction
        Return New Fraction(BigInteger.Multiply(Fraction1.Numerator, Fraction2.Numerator), BigInteger.Multiply(Fraction1.Denominator, Fraction2.Denominator))
    End Operator
    Public Shared Operator /(Fraction1 As Fraction, Fraction2 As Fraction) As Fraction
        Return New Fraction(BigInteger.Multiply(Fraction1.Numerator, Fraction2.Denominator), BigInteger.Multiply(Fraction1.Denominator, Fraction2.Numerator))
    End Operator
    Public Shared Operator ^(Fraction As Fraction, Power As Long) As Fraction
        If Power > 0 Then
            Return New Fraction(BigInteger.Pow(Fraction.Numerator, Power), BigInteger.Pow(Fraction.Denominator, Power))
        ElseIf Power < 0 Then
            Return New Fraction(BigInteger.Pow(Fraction.Denominator, Power), BigInteger.Pow(Fraction.Numerator, Power))
        End If
        Return New Fraction(1, 1)
    End Operator
    Public Shared Operator -(Fraction As Fraction) As Fraction
        Return 0 - Fraction
    End Operator

    Public Shared Operator =(Fraction1 As Fraction, Fraction2 As Fraction) As Boolean
        If Fraction1.IsNull And Fraction2.IsNull Then Return True
        If (Fraction1 / Fraction2).IsUnit Then Return True
        Return False
    End Operator
    Public Shared Operator <>(Fraction1 As Fraction, Fraction2 As Fraction) As Boolean
        If Fraction1 = Fraction2 Then Return False
        Return True
    End Operator
    Public Shared Operator >(Fraction1 As Fraction, Fraction2 As Fraction) As Boolean
        If Fraction2.IsNull Then
            If Fraction1.IsNegitive Then
                Return False
            Else
                Return True
            End If
        End If
        Dim Division = Fraction1 / Fraction2
        If Division.Numerator > Division.Denominator Then Return True
        Return False
    End Operator
    Public Shared Operator <(Fraction1 As Fraction, Fraction2 As Fraction) As Boolean
        If Fraction2.IsNull Then
            If Fraction1.IsNegitive Then
                Return True
            Else
                Return False
            End If
        End If
        Dim Division = Fraction1 / Fraction2
        If Division.Numerator < Division.Denominator Then Return True
        Return False
    End Operator
    Public Shared Operator >=(Fraction1 As Fraction, Fraction2 As Fraction) As Boolean
        If Fraction1 > Fraction2 Or Fraction1 = Fraction2 Then Return True
        Return False
    End Operator
    Public Shared Operator <=(Fraction1 As Fraction, Fraction2 As Fraction) As Boolean
        If Fraction1 < Fraction2 Or Fraction1 = Fraction2 Then Return True
        Return False
    End Operator

    Public Shared Widening Operator CType(Number As Byte) As Fraction
        Return New Fraction(Number, 1)
    End Operator
    Public Shared Widening Operator CType(Number As SByte) As Fraction
        Return New Fraction(Number, 1)
    End Operator
    Public Shared Widening Operator CType(Number As Short) As Fraction
        Return New Fraction(Number, 1)
    End Operator
    Public Shared Widening Operator CType(Number As UShort) As Fraction
        Return New Fraction(Number, 1)
    End Operator
    Public Shared Widening Operator CType(Number As Integer) As Fraction
        Return New Fraction(Number, 1)
    End Operator
    Public Shared Widening Operator CType(Number As UInteger) As Fraction
        Return New Fraction(Number, 1)
    End Operator
    Public Shared Widening Operator CType(Number As Long) As Fraction
        Return New Fraction(Number, 1)
    End Operator
    Public Shared Widening Operator CType(Number As ULong) As Fraction
        Return New Fraction(Number, 1)
    End Operator
    Public Shared Widening Operator CType(Number As BigInteger) As Fraction
        Return New Fraction(Number, 1)
    End Operator
    Public Shared Widening Operator CType(Number As Single) As Fraction
        Return Fraction.Parse(Number.ToString)
    End Operator
    Public Shared Widening Operator CType(Number As Double) As Fraction
        Return Fraction.Parse(Number.ToString)
    End Operator
    Public Shared Widening Operator CType(Number As Decimal) As Fraction
        Return Fraction.Parse(Number.ToString)
    End Operator

    Public Shared Widening Operator CType(Fraction As Fraction) As Byte
        Return Fraction.WholeNumber
    End Operator
    Public Shared Widening Operator CType(Fraction As Fraction) As SByte
        Return Fraction.WholeNumber
    End Operator
    Public Shared Widening Operator CType(Fraction As Fraction) As Short
        Return Fraction.WholeNumber
    End Operator
    Public Shared Widening Operator CType(Fraction As Fraction) As UShort
        Return Fraction.WholeNumber
    End Operator
    Public Shared Widening Operator CType(Fraction As Fraction) As Integer
        Return Fraction.WholeNumber
    End Operator
    Public Shared Widening Operator CType(Fraction As Fraction) As UInteger
        Return Fraction.WholeNumber
    End Operator
    Public Shared Widening Operator CType(Fraction As Fraction) As Long
        Return Fraction.WholeNumber
    End Operator
    Public Shared Widening Operator CType(Fraction As Fraction) As ULong
        Return Fraction.WholeNumber
    End Operator
    Public Shared Widening Operator CType(Fraction As Fraction) As Single
        Return Single.Parse(Fraction.ToString(10))
    End Operator
    Public Shared Widening Operator CType(Fraction As Fraction) As Double
        Return Single.Parse(Fraction.ToString(20))
    End Operator

    Public Shared ReadOnly Property Unit As Fraction
        Get
            Return New Fraction(1, 1)
        End Get
    End Property
    Public Shared ReadOnly Property Null As Fraction
        Get
            Return New Fraction(0, 1)
        End Get
    End Property
    Public Shared ReadOnly Property Undefined As Fraction
        Get
            Return New Fraction(0, 0)
        End Get
    End Property
    Public Shared Function GCD(First As BigInteger, Second As BigInteger) As BigInteger
        First = BigInteger.Abs(First)
        Second = BigInteger.Abs(Second)
        Dim Reminder As BigInteger
        While Second > 0
            Reminder = First Mod Second
            First = Second
            Second = Reminder
        End While
        Return First
    End Function
    Public Shared Function TryParse(Value As String, ByRef Result As Fraction) As Boolean
        Value = Value.Trim(" ")
        If Value.Contains("/") Then
            Dim Array() = Value.Split("/")
            If Array.Length <> 2 Then Return False
            Dim IsConverted As Boolean
            Dim Numerator As BigInteger = 0
            Dim Denominator As BigInteger = 0
            IsConverted = BigInteger.TryParse(Array(0), Numerator)
            If IsConverted = False Then Return False
            IsConverted = BigInteger.TryParse(Array(1), Denominator)
            If IsConverted = False Then Return False
            Result = New Fraction(Numerator, Denominator)
        ElseIf Value.Contains(".") Then
            Dim IsNegitive As Boolean = False
            If Value(0) = "-" Then IsNegitive = True
            Dim Array() = Value.Split(".")
            If Array.Length <> 2 Then Return False
            If Array(1).Contains("-") Then Return False
            Dim IsConverted As Boolean
            Dim WholeNumber As BigInteger = 0
            If Array(0) <> "" Then
                IsConverted = BigInteger.TryParse(Array(0), WholeNumber)
                If IsConverted = False Then Return False
                WholeNumber = BigInteger.Abs(WholeNumber)
            End If
            Dim ProperNumerator As BigInteger = 0
            IsConverted = BigInteger.TryParse(Array(1), ProperNumerator)
            If IsConverted = False Then Return False
            Dim ProperDenominator(Array(1).Length) As Char
            ProperDenominator(0) = "1"c
            For i = 1 To Array(1).Length
                ProperDenominator(i) = "0"c
            Next
            Dim Proper As New Fraction(ProperNumerator, BigInteger.Parse(ProperDenominator))
            If IsNegitive Then
                Result = -New Fraction(WholeNumber, Proper)
            Else
                Result = New Fraction(WholeNumber, Proper)
            End If
        Else
            Dim IsConverted As Boolean
            Dim Number As BigInteger = 0
            IsConverted = BigInteger.TryParse(Value, Number)
            If IsConverted = False Then Return False
            Result = New Fraction(Number)
        End If
        Return True
    End Function
    Public Shared Function Parse(Value As String) As Fraction
        Dim Fraction As Fraction = 0
        Dim IsConverted As Boolean
        IsConverted = TryParse(Value, Fraction)
        If IsConverted = False Then
            Throw New Exception("Invalid format. Unable to convert " + Value.ToString + " to fraction.")
        Else
            Return Fraction
        End If
    End Function
End Structure

File: Fraction_cs/Fraction.cs
using System;
using System.Numerics;
using System.Text;
public struct Fraction : IComparable, IComparable<Fraction>, IEquatable<Fraction>
{
    private BigInteger _Numerator;
    private BigInteger _Denominator;
    public BigInteger Numerator
    {
        get { return _Numerator; }
        set
        {
            _Numerator = value;
            Simplify();
        }
    }
    public BigInteger Denominator
    {
        get { return _Denominator; }
        set
        {
            _Denominator = value;
            Simplify();
        }
    }
    public BigInteger WholeNumber
    {
        get
        {
            if (Denominator == 0)
                return 0;
            return BigInteger.Divide(Numerator, Denominator);
        }
        set
        {
            Fraction pf = ProperFraction;
            Fraction Fraction = pf + value;
            _Numerator = Fraction.Numerator;
            _Denominator = Fraction.Denominator;
            Simplify();
        }
    }
    public Fraction ProperFraction
    {
        get
        {
            if (Denominator == 0)
                return new Fraction(0, 0);
            return new Fraction(Numerator % Denominator, Denominator);
        }
        set
        {
            BigInteger  wn = WholeNumber;
            Fraction Fraction = wn + value;
            _Numerator = Fraction.Numerator;
            _Denominator = Fraction.Denominator;
            Simplify();
        }
    }
    public bool IsProper
    {
        get { return BigInteger.Abs(Numerator) < Denominator; }
    }
    public bool IsImproper
    {
        get { return !IsProper; }
    }
    public bool IsNull
    {
        get { return Numerator == 0; }
    }
    public bool IsUnit
    {
        get
        {
            return Numerator == 1 && Denominator == 1;
        }
    }
    public bool IsNegitive
    {
        get { return Numerator < 0; }
    }
    public bool IsUndefined
    {
        get { return Denominator == 0; }
    }
    public Fraction Absolute
    {
        get { return new Fraction(BigInteger.Abs(Numerator), BigInteger.Abs(Denominator)); }
    }

    public new string ToString()
    {
        return _Numerator.ToString() + "/" + _Denominator.ToString();
    }
    public string ToString(ulong DigitsAfterPoint)
    {
        StringBuilder sb = new StringBuilder();
        BigInteger Whole = BigInteger.Abs(WholeNumber);
        Fraction Proper = ProperFraction.Absolute;
        sb.AppendFormat("{0}.", Whole.ToString());
        BigInteger Numerator = Proper.Numerator;
        BigInteger Denominator = Proper.Denominator;
        for (ulong i = 0; i <= DigitsAfterPoint - 1; i++)
        {
            Numerator = Numerator * 10;
            sb.Append(BigInteger.Divide(Numerator, Denominator).ToString());
            Numerator = Numerator % Denominator;
        }
        if (IsNegitive)
            sb.Insert(0, "-");
        return sb.ToString();
    }

    public void SetValue(BigInteger Numerator, BigInteger Denominator)
    {
        _Numerator = Numerator;
        _Denominator = Denominator;
        Simplify();
    }
    public void Simplify()
    {
        if (Denominator == 0)
        {
            _Numerator = 0;
            _Denominator = 0;
        }
        else if (Denominator < 0)
        {
            _Denominator = BigInteger.Abs(Denominator);
            _Numerator = -Numerator;
        }
        else if (Numerator == 0 & Denominator != 0)
        {
            _Denominator = 1;
        }
        else if (Numerator == Denominator & Denominator != 0)
        {
            _Numerator = 1;
            _Denominator = 1;
        }
        else {
            BigInteger GDC = Fraction.GCD(Numerator, Denominator);
            _Numerator = Numerator / GDC;
            _Denominator = Denominator / GDC;
        }
    }
    public void Reduce(long Change = 1)
    {
        Reduce(Change, Change);
    }
    public void Reduce(long ChangeInNumerator, long ChangeInDenominator)
    {
        bool IsNegitive = false;
        Fraction Proper = ProperFraction;
        if (Proper < 0)
        {
            IsNegitive = true;
            Proper = Proper.Absolute;
        }
        Fraction Reduced = Proper;
        for (long CIN = -ChangeInNumerator; CIN <= ChangeInNumerator; CIN++)
        {
            for (long CID = -ChangeInDenominator; CID <= ChangeInDenominator; CID++)
            {
                Fraction Fraction = (new Fraction(Proper.Numerator + CIN, Proper.Denominator + CID)).Absolute;
                if (Fraction.Numerator < Reduced.Numerator && Fraction.Denominator < Reduced.Denominator && !Fraction.IsUndefined)
                    Reduced = Fraction;
            }
        }
        if (IsNegitive == true)
        {
            ProperFraction = -Reduced;
        }
        else {
            ProperFraction = Reduced;
        }
    }
    public static Fraction Power(Fraction Fraction, int Power)
    {
        if (Power > 0)
        {
            return new Fraction(BigInteger.Pow(Fraction.Numerator, Power), BigInteger.Pow(Fraction.Denominator, Power));
        }
        else if (Power < 0)
        {
            return new Fraction(BigInteger.Pow(Fraction.Denominator, Power), BigInteger.Pow(Fraction.Numerator, Power));
        }
        return new Fraction(1, 1);
    }

    public int CompareTo(object obj)
    {
        if (obj == null)
            return 1;
        if (obj is Fraction)
        {
            return CompareTo((Fraction)obj);
        }
        else {
            throw new ArgumentException(string.Format("{0} is not a {1}", nameof(obj), GetType().FullName));
        }
    }
    public int CompareTo(Fraction other)
    {
        if (this == other)
            return 0;
        if (this > other)
        {
            return 1;
        }
        else {
            return -1;
        }
    }
    public bool Equals(Fraction other)
    {
        return this == other;
    }

    public Fraction(BigInteger Numerator, BigInteger Denominator)
    {
        _Numerator = Numerator;
        _Denominator = Denominator;
        Simplify();
    }
    public Fraction(BigInteger Number)
    {
        _Numerator = Number;
        _Denominator = 1;
    }
    public Fraction(BigInteger WholeNumber, Fraction ProperFraction)
    {
        _Numerator = WholeNumber * ProperFraction.Denominator + ProperFraction.Numerator;
        _Denominator = ProperFraction.Denominator;
    }
    public Fraction(string Value)
    {
        Fraction Fraction = Parse(Value);
        _Numerator = Fraction.Numerator;
        _Denominator = Fraction.Denominator;
        Simplify();
    }
    public Fraction(decimal Number) : this(Number.ToString())
    {
    }
    public Fraction(double Number) : this(Number.ToString())
    {
    }
    public Fraction(float Number) : this(Number.ToString())
    {
    }

    public static Fraction operator +(Fraction Fraction1, Fraction Fraction2)
    {
        return new Fraction(BigInteger.Add(BigInteger.Multiply(Fraction1.Numerator, Fraction2.Denominator), BigInteger.Multiply(Fraction2.Numerator, Fraction1.Denominator)), BigInteger.Multiply(Fraction1.Denominator, Fraction2.Denominator));
    }
    public static Fraction operator -(Fraction Fraction1, Fraction Fraction2)
    {
        return new Fraction(BigInteger.Subtract(BigInteger.Multiply(Fraction1.Numerator, Fraction2.Denominator), BigInteger.Multiply(Fraction2.Numerator, Fraction1.Denominator)), BigInteger.Multiply(Fraction1.Denominator, Fraction2.Denominator));
    }
    public static Fraction operator *(Fraction Fraction1, Fraction Fraction2)
    {
        return new Fraction(BigInteger.Multiply(Fraction1.Numerator, Fraction2.Numerator), BigInteger.Multiply(Fraction1.Denominator, Fraction2.Denominator));
    }
    public static Fraction operator /(Fraction Fraction1, Fraction Fraction2)
    {
        return new Fraction(BigInteger.Multiply(Fraction1.Numerator, Fraction2.Denominator), BigInteger.Multiply(Fraction1.Denominator, Fraction2.Numerator));
    }
    public static Fraction operator -(Fraction Fraction)
    {
        return 0 - Fraction;
    }

    public static bool operator ==(Fraction Fraction1, Fraction Fraction2)
    {
        if (Fraction1.IsNull & Fraction2.IsNull)
            return true;
        if ((Fraction1 / Fraction2).IsUnit)
            return true;
        return false;
    }
    public static bool operator !=(Fraction Fraction1, Fraction Fraction2)
    {
        if (Fraction1 == Fraction2)
            return false;
        return true;
    }
    public static bool operator >(Fraction Fraction1, Fraction Fraction2)
    {
        if (Fraction2.IsNull)
        {
            if (Fraction1.IsNegitive)
            {
                return false;
            }
            else {
                return true;
            }
        }
        Fraction Division = Fraction1 / Fraction2;
        if (Division.Numerator > Division.Denominator)
            return true;
        return false;
    }
    public static bool operator <(Fraction Fraction1, Fraction Fraction2)
    {
        if (Fraction2.IsNull)
        {
            if (Fraction1.IsNegitive)
            {
                return true;
            }
            else {
                return false;
            }
        }
        Fraction Division = Fraction1 / Fraction2;
        if (Division.Numerator < Division.Denominator)
            return true;
        return false;
    }
    public static bool operator >=(Fraction Fraction1, Fraction Fraction2)
    {
        if (Fraction1 > Fraction2 | Fraction1 == Fraction2)
            return true;
        return false;
    }
    public static bool operator <=(Fraction Fraction1, Fraction Fraction2)
    {
        if (Fraction1 < Fraction2 | Fraction1 == Fraction2)
            return true;
        return false;
    }

    public static implicit operator Fraction(byte Number)
    {
        return new Fraction(Number, 1);
    }
    public static implicit operator Fraction(sbyte Number)
    {
        return new Fraction(Number, 1);
    }
    public static implicit operator Fraction(short Number)
    {
        return new Fraction(Number, 1);
    }
    public static implicit operator Fraction(ushort Number)
    {
        return new Fraction(Number, 1);
    }
    public static implicit operator Fraction(int Number)
    {
        return new Fraction(Number, 1);
    }
    public static implicit operator Fraction(uint Number)
    {
        return new Fraction(Number, 1);
    }
    public static implicit operator Fraction(long Number)
    {
        return new Fraction(Number, 1);
    }
    public static implicit operator Fraction(ulong Number)
    {
        return new Fraction(Number, 1);
    }
    public static implicit operator Fraction(BigInteger Number)
    {
        return new Fraction(Number, 1);
    }
    public static implicit operator Fraction(float Number)
    {
        return Fraction.Parse(Number.ToString());
    }
    public static implicit operator Fraction(double Number)
    {
        return Fraction.Parse(Number.ToString());
    }
    public static implicit operator Fraction(decimal Number)
    {
        return Fraction.Parse(Number.ToString());
    }

    public static explicit operator byte(Fraction Fraction)
    {
        return Convert.ToByte(Fraction.WholeNumber);
    }
    public static explicit operator sbyte(Fraction Fraction)
    {
        return Convert.ToSByte(Fraction.WholeNumber);
    }
    public static explicit operator short(Fraction Fraction)
    {
        return Convert.ToInt16(Fraction.WholeNumber);
    }
    public static explicit operator ushort(Fraction Fraction)
    {
        return Convert.ToUInt16(Fraction.WholeNumber);
    }
    public static explicit operator int(Fraction Fraction)
    {
        return Convert.ToInt32(Fraction.WholeNumber);
    }
    public static explicit operator uint(Fraction Fraction)
    {
        return Convert.ToUInt32(Fraction.WholeNumber);
    }
    public static explicit operator long(Fraction Fraction)
    {
        return Convert.ToInt64(Fraction.WholeNumber);
    }
    public static explicit operator ulong(Fraction Fraction)
    {
        return Convert.ToUInt64(Fraction.WholeNumber);
    }
    public static explicit operator float(Fraction Fraction)
    {
        return float.Parse(Fraction.ToString(10));
    }
    public static explicit operator double(Fraction Fraction)
    {
        return float.Parse(Fraction.ToString(20));
    }

    public static Fraction Unit
    {
        get { return new Fraction(1, 1); }
    }
    public static Fraction Null
    {
        get { return new Fraction(0, 1); }
    }
    public static Fraction Undefined
    {
        get { return new Fraction(0, 0); }
    }
    public static BigInteger GCD(BigInteger First, BigInteger Second)
    {
        First = BigInteger.Abs(First);
        Second = BigInteger.Abs(Second);
        BigInteger Reminder = default(BigInteger);
        while (Second > 0)
        {
            Reminder = First % Second;
            First = Second;
            Second = Reminder;
        }
        return First;
    }
    public static bool TryParse(string Value, ref Fraction Result)
    {
        if (Value == null || string.IsNullOrEmpty(Value))
            return false;
        Value = Value.Trim(' ');
        bool IsConverted = false;
        if (Value.Contains("/"))
        { 
            string[] tmp = Value.Split('/');
			if (tmp.Length != 2)
				return false;
			BigInteger Numerator = 0;
            BigInteger Denominator = 0;
            IsConverted = BigInteger.TryParse(tmp[0],out Numerator);
			if (IsConverted == false)
				return false;
			IsConverted = BigInteger.TryParse(tmp[1],out Denominator);
			if (IsConverted == false)
				return false;
			Result = new Fraction(Numerator, Denominator);
        } else if (Value.Contains(".")) {
			bool IsNegitive = false;
			if (Value[0]== "-"[0])
				IsNegitive = true;
                string[] tmp = Value.Split('.');
			if (tmp.Length != 2)
				return false;
			if (tmp[1].Contains("-"))
				return false;
			BigInteger WholeNumber = 0;
			if (!string.IsNullOrEmpty(tmp[0])) {
				IsConverted = BigInteger.TryParse(tmp[0],out WholeNumber);
				if (IsConverted == false)
					return false;
				WholeNumber = BigInteger.Abs(WholeNumber);
			}
			BigInteger ProperNumerator = 0;
            IsConverted = BigInteger.TryParse(tmp[1],out ProperNumerator);
			if (IsConverted == false)
				return false;
			char[] ProperDenominator = new char[tmp[1].Length + 1];
            ProperDenominator[0] = '1';
			for (int i = 1; i <= tmp[1].Length; i++) {
                ProperDenominator[i] = '0';
			}
			Fraction Proper = new Fraction(ProperNumerator, BigInteger.Parse(new string(ProperDenominator)));
			if (IsNegitive) {
				Result = -new Fraction(WholeNumber, Proper);
			} else {
				Result = new Fraction(WholeNumber, Proper);
			}
		} else {
			BigInteger Number = 0;
            IsConverted = BigInteger.TryParse(Value,out Number);
			if (IsConverted == false)
				return false;
			Result = new Fraction(Number);
		}
		return true;
	}
	public static Fraction Parse(string Value)
{
    Fraction Fraction = 0;
    bool IsConverted = false;
    IsConverted = TryParse(Value, ref Fraction);
    if (IsConverted == false)
    {
        throw new FormatException("Invalid format. Unable to convert " + Value.ToString() + " to fraction.");
    }
    else {
        return Fraction;
    }
}
}